jetbrains-core/src/software/aws/toolkits/jetbrains/services/ecr/actions/PushTagToRepositoryAction.kt

```diff
@@ -0,0 +1,174 @@
+            RemoteServerImpl("DockerConnection", DockerCloudType.getInstance(), DockerCloudConfiguration.createDefault())
+        )
+        connection.connectIfNeeded(object : ServerConnector.ConnectionCallback<DockerDeploymentConfiguration> {
+            override fun errorOccurred(errorMessage: String) {
+                instancePromise.setError(errorMessage)
+            }
+
+            override fun connected(serverRuntimeInstance: ServerRuntimeInstance<DockerDeploymentConfiguration>) {
+                instancePromise.setResult(serverRuntimeInstance as DockerServerRuntimeInstance)
+            }
+        })
+
+        return instancePromise.await()
+    }
+
+    private suspend fun getDockerApplicationRuntimeInstance(serverRuntime: DockerServerRuntimeInstance, imageId: String): DockerApplicationRuntime =
+        serverRuntime.findRuntimeLater(imageId, false).await()
+}
+
+internal data class LocalImage(
+    val imageId: String,
+    val tag: String?
+)
+
+internal class PushToEcrDialog(
+    project: Project,
+    private val dockerServerRuntime: Deferred<DockerServerRuntimeInstance>
+) : DialogWrapper(project, null, false, IdeModalityType.PROJECT),
+    CoroutineScope by ApplicationThreadPoolScope("PushRepositoryDialog") {
+    private var localImage: LocalImage? = null
+    private var remoteTag: String = "latest"
+    private val localImageRepoTags = CollectionComboBoxModel<LocalImage>()
+
+    private val remoteRepos = ResourceSelector.builder()
+        .resource(EcrResources.LIST_REPOS)
+        .customRenderer(SimpleListCellRenderer.create("") { it.repositoryName })
+        .awsConnection(project)
+        .build()
+
+    init {
+        title = message("ecr.push.title")
+        setOKButtonText(message("ecr.push.confirm"))
+
+        init()
+
+        launch {
+            val serverRuntime = dockerServerRuntime.await()
+            val images = serverRuntime.agent.getImages(null)
+            localImageRepoTags.add(images.toLocalImageList())
+            localImageRepoTags.update()
+        }
+    }
+
+    override fun createCenterPanel() = panel {
+        row(message("ecr.repo.label")) {
+            component(remoteRepos)
+        }
+
+        row(message("ecr.push.source")) {
+            // property binding syntax causes kotlin compiler error for some reason
+            comboBox(
+                localImageRepoTags,
+                { localImage },
+                { ::localImage.set(it) },
+                listCellRenderer { value, _, _ ->
+                    value.tag ?: value.imageId.take(15)
+                }
+            ).withErrorOnApplyIf(message("ecr.repo.not_selected")) { it.selected() == null }
+        }
+
+        row(message("ecr.push.remoteTag")) {
+            textField(::remoteTag)
+                .withErrorOnApplyIf("local image not specified") { it.blankAsNull() == null }
+        }
+    }
+
+    fun getPushRequest() = EcrPushRequest(
```

Comment:
Need to add `doValidate` or `validateAll` to the panel

Reply:
`withErrorOnApplyIf` takes the place of `doValidate` if that's what you mean.

(`doValidate` appears to run before values are committed back to the underlying fields, so `doValidate` will always see the default values and return failure)